rskj-core/src/main/java/org/ethereum/datasource/TransientMap.java

```diff
@@ -0,0 +1,212 @@
+        }
+
+        public V get(Object key) {
+            if (m==null)
+                return added.get(key);
+
+            if (deleted.contains(key))
+                return null;
+            V ret = added.get(key);
+            if (ret==null)
+                ret = changed.get(key);
+            if (ret==null)
+                ret = this.m.get(key);
+            return ret;
+        }
+
+        public V put(K key, V value) {
+            if (m==null)
+                return added.put(key,value);
+
+            V ret = m.get(key);
+            if (ret!=null) {
+                if (deleted.contains(key)) {
+                    deleted.remove(key);
+                    changed.put(key,value);
+                    return null;
+                }
+                return changed.put(key,value);
+            } else
+                return added.put(key,value);
+        }
+
+        public V remove(Object key) {
+            V ret;
+            if (m!=null) {
+                if (!deleted.contains(key)) {
+                    deleted.add((K) key);
+                    ret = changed.remove(key);
+                    if (ret==null)
+                        ret = added.remove(key);
+                    if (ret==null)
+                        ret = m.get(key);
+                } else
+                    // double delete
+                    ret = null;
+            } else {
+                ret = added.remove(key);
+            }
+            return ret;
+        }
+
+        public void putAll(Map<? extends K, ? extends V> am) {
+
+            if (m==null) {
+                return;
+            }
+            for (Map.Entry<? extends K,? extends V> entry : am.entrySet()) {
+                deleted.remove(entry.getKey());
+                if (m.containsKey(entry.getKey())) {
+                    changed.put(entry.getKey(),entry.getValue());
+                } else
+                    added.put(entry.getKey(),entry.getValue());
+            }
+        }
+
+        public void clear() {
+            m = null;
+            deleted.clear();
+            added.clear();
+            changed.clear();
+        }
+
+        public Set<K> keySet() {
+            if (this.keySet == null) {
+                //TODO:  this.keySet = Collections.unmodifiableSet(this.m.keySet());
+            }
+
+            return this.keySet;
+        }
+
+        public Set<Entry<K, V>> entrySet() {
+            if (this.entrySet == null) {
+                //TODO: this.entrySet = unmodifiableEntrySet(this.m.entrySet());
+            }
+
+            return this.entrySet;
+        }
+
+        public Collection<V> values() {
+            if (this.values == null) {
+                //TODO:  this.values = Collections.unmodifiableCollection(this.m.values());
+            }
+
+            return this.values;
+        }
+
+
+        public V getOrDefault(Object k, V defaultValue) {
+            V ret = get(k);
+            if (k==null)
```

Comment:
Shouldn't it be `ret == null`?

Reply:
I've implemented this, please open a new thread if you think we shouldn't @SergioDemianLerner.